launchbear.1
.TH LAUNCHBEAR 1 "2018-2025" launchbear "User Commands"
.SH NAME
launchbear - extensible program launcher
.SH SYNOPSIS
.B launchbear
[options...]
.SH DESCRIPTION
This command launches programs.
.SH OPTIONS
.TP
.BR \-d ", " \-\-backend-dir
Alternative location for backend scripts
.TP
.BR \-v ", " \-\-verbose
Emit debugging output
.SH AUTHOR
Written and maintained by P. S. Clarke


debian/changelog
launchbear (2.3.6) unstable; urgency=medium

  * Add -v for verbose mode.
  * Backends: Aisleriot: Handle when not installed.

 -- P. S. Clarke <debs@pscl4rke.net>  Wed, 22 Oct 2025 18:17:43 +0100

launchbear (2.3.5) unstable; urgency=medium

  * Backends: XDG: Ignore when NoDisplay=true.

 -- P. S. Clarke <debs@pscl4rke.net>  Fri, 16 May 2025 17:46:06 +0100

launchbear (2.3.4) unstable; urgency=medium

  * Backends: XDG: Use $XDG_DATA_DIRS.

 -- P. S. Clarke <debs@pscl4rke.net>  Mon, 10 Mar 2025 21:18:05 +0000

launchbear (2.3.3) unstable; urgency=medium

  * Wizard: Test out whether scripts are runnable.
  * Backends: Quicker to exit with errors if not valid.

 -- P. S. Clarke <debs@pscl4rke.net>  Wed, 17 Apr 2024 13:48:21 +0100

launchbear (2.3.2) unstable; urgency=medium

  * Crude Wofi Support

 -- P. S. Clarke <debs@pscl4rke.net>  Sun, 30 Oct 2022 13:38:34 +0000

launchbear (2.3.1) unstable; urgency=medium

  * Cache: Fix bug that stopped it from working.
  * Backends: Aisleriot: Fix speed issue.

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 01 Jan 2022 22:25:13 +0000

launchbear (2.3.0) unstable; urgency=medium

  * Backends: XDG: Ignore XScreenSaver files.
  * Add a -d flag to customise backend directory.
  * Backends: Aisleriot: New backend for card games.

 -- P. S. Clarke <debs@pscl4rke.net>  Sun, 12 Sep 2021 21:49:04 +0100

launchbear (2.2.1) unstable; urgency=medium

  * Backends: XDG: Fix to be more resilient.

 -- P. S. Clarke <debs@pscl4rke.net>  Mon, 06 Jul 2020 19:39:24 +0100

launchbear (2.2.0) unstable; urgency=medium

  * Backends: Add linkbear and remove srprompt.

 -- P. S. Clarke <debs@pscl4rke.net>  Wed, 22 Apr 2020 18:18:39 +0100

launchbear (2.1.0) unstable; urgency=medium

  * Wizard: Detect if symlinks already exist.
  * Wizard: Only delete cache file if necessary.
  * Wizard: Add 'p' option to preview backend output.

 -- P. S. Clarke <debs@pscl4rke.net>  Wed, 25 Mar 2020 08:33:59 +0000

launchbear (2.0.0) unstable; urgency=medium

  * Switch from Python 2 to Python 3.

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 09 Nov 2019 14:50:38 +0000

launchbear (1.0.4) unstable; urgency=medium

  * Use HTTPS for home page URLs.
  * Add .py file extensions.
  * Introduce a placeholder test suite.
  * Fix bug with corrupted cache file.

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 09 Nov 2019 10:12:13 +0000

launchbear (1.0.3) unstable; urgency=medium

  * Fix python2 dependency.

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 10 Mar 2018 18:08:52 +0000

launchbear (1.0.2) unstable; urgency=medium

  * Add README and license with the GPLv3.
  * Add manpages for the binaries.

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 10 Mar 2018 17:38:23 +0000

launchbear (1.0.1) unstable; urgency=medium

  * Fix dependency from python2 to python2.7

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 06 Jan 2018 20:30:09 +0000

launchbear (1.0.0) unstable; urgency=medium

  * Initial Release.

 -- P. S. Clarke <debs@pscl4rke.net>  Sat, 06 Jan 2018 20:18:33 +0000


launchbear.py
#!/usr/bin/python3

"""
//  LAUNCHBEAR

Launchbear is a piece of middleware designed to launch applications
in a GUI environment.  It takes selectable items from plugins, and
allows them to be picked using a front-end.
"""


import logging
LOG = logging.getLogger("launchbear")

import argparse
import os
import shlex
import subprocess
import pickle
import sys


class HistoryFile:
    """Tracks what has been launched to order by popularity."""

    max_history = 500

    def __init__(self):
        self.location = os.path.join(os.environ['HOME'], ".launchbear", "history")
        self.history = []
        try:
            LOG.info("Loading history from: %r", self.location)
            with open(self.location) as infile:
                for line in infile:
                    self.history.append(line.strip())
        except IOError as exc:
            # Presumably doesn't exist yet
            LOG.warning("Loading history failed: %r", exc)

    def update(self, choice_id):
        self.history.append(choice_id)
        if len(self.history) > self.max_history:
            self.history.pop(0)
        with open(self.location, 'w') as outfile:
            for choice_id in self.history:
                outfile.write("%s\n" % choice_id)

    def score(self, choice_id):
        score = 0
        for previous in self.history:
            if previous == choice_id:
                score += 1
        return score


class ChoiceGenerator:
    """A backend that provides a number of choices."""

    def __init__(self, backend_location):
        """Create a new ChoiceGenerator."""
        self._backend_location = backend_location
        self._choices = []

    def choices(self):
        """Returns a set of choices."""
        choices_available = {}
        for choice in self._choices:
            if "id" not in choice:
                continue
            choices_available[choice["id"]] = choice
        return choices_available

    def run_until_loaded(self):
        # these two parts could be separated so that each Popen is
        # run in parallel for improved responsiveness
        self.process = subprocess.Popen(self._backend_location,
                    stdout=subprocess.PIPE, shell=True)
        self.parse(self.process.stdout)

    def parse(self, stream):
        """Build a set of choices from a file stream."""
        for line in stream.readlines():
            arguments = shlex.split(line.decode(), comments=True)
            if len(arguments) == 0:
                continue
            handler_name = "handle_%s" % arguments[0]
            if not hasattr(self, handler_name):
                print("Unknown directive '%s'" % arguments[0])
            else:
                getattr(self, handler_name)(arguments[1:])

    def load(self, cached_data):
        """Build a set of choices from previously cached data."""
        if 'choices' in cached_data:
            self._choices = cached_data['choices']
        else:
            self._choices = {}

    def save(self):
        """Returns the cache form that can be loaded."""
        return {
            'choices': self._choices,
        }

    def handle_addchoice(self, arguments):
        choice_properties = {}
        for argument in arguments:
            if not argument.startswith("--"):
                continue
            dashed_key, equals, value = argument.partition("=")
            if equals != "=":
                continue
            choice_properties[dashed_key[2:]] = value
        self._choices.append(choice_properties)


class DmenuFrontend:
    """A picker based around dmenu."""

    def command(self):  # Gets overridden
        return [
            "dmenu", "-i", "-l", "6",
            "-fn", "-*-arial-*-r-*-*-*-*-*-*-*-*-*",
            "-nb", "#444", "-nf", "#DDD", "-sb", "#757",
            "-p", "Run:",
        ]

    def pick(self, choices):
        """Presents the choices and returns the id chosen or None."""
        lines = []
        for choice in choices:
            parts = []
            if 'title' in choice:
                parts.append(choice['title'])
            if 'cmd' not in choice:
                continue
            parts.append("(%s)" % choice['cmd'])
            if 'id' not in choice:
                continue
            parts.append("||| %s" % choice['id'])
            lines.append(" ".join(parts))
        command = self.command()
        process = subprocess.Popen(command,
                    stdin=subprocess.PIPE, stdout=subprocess.PIPE)
        stdout, stderr = process.communicate(("\n".join(lines)).encode())
        if len(stdout) == 0:
            return None
        rest, bars, choice_id = stdout.strip().decode().partition("||| ")
        return choice_id


class WofiFrontend(DmenuFrontend):

    def command(self):  # Overriding
        return ["wofi", "-d", "-i"]


def decide_on_frontend():
    # FIXME This is a terrible way of doing things!
    if os.path.exists("/usr/bin/wofi"):
        return WofiFrontend()
    return DmenuFrontend()


def default_cache_path():
    homedir = os.environ['HOME']
    return os.path.join(homedir, ".launchbear/cache.pkl")


def load_cachefile(cache_path=None):
    """Return the cached data, or blank data."""
    if cache_path is None:
        cache_path = default_cache_path()
    try:
        LOG.info("Loading cache from: %r", cache_path)
        with open(cache_path, "rb") as cache_file:
            return pickle.load(cache_file)
    except Exception as exc:
        LOG.warning("Loading cache failed: %r", exc)
        return {'generators': {}}


def save_cachefile(cache, cache_path=None):
    """Saves the given cache to disk."""
    if cache_path is None:
        cache_path = default_cache_path()
    with open(cache_path, "wb") as cache_file:
        pickle.dump(cache, cache_file)


def parse_args(args):
    parser = argparse.ArgumentParser()
    parser.add_argument("-d", "--backend-dir", default="~/.launchbear/backends")
    parser.add_argument("-v", "--verbose", action="store_const", dest="logging_level",
                        const="DEBUG", default="WARNING")
    return parser.parse_args(args)


def main():
    opts = parse_args(sys.argv[1:])
    logging.basicConfig(level=opts.logging_level)
    backend_names = os.listdir(os.path.expanduser(opts.backend_dir))
    cache = load_cachefile()
    history = HistoryFile()
    all_choices = {}
    for backend_name in backend_names:
        backend_location = os.path.join(opts.backend_dir, backend_name)
        generator = ChoiceGenerator(backend_location)
        if backend_name in cache['generators']:
            LOG.info("Loading from cache: %r", backend_name)
            generator.load(cache['generators'][backend_name])
        else:
            LOG.info("Running: %r", backend_name)
            generator.run_until_loaded()
            cache['generators'][backend_name] = generator.save()
        choices_available = generator.choices()
        all_choices.update(choices_available)
    save_cachefile(cache)
    frontend = decide_on_frontend()
    #sortable_choices = [(history.score(x['id']), x) for x in all_choices.values()]
    #sortable_choices.sort(reverse=True, key=lambda (a, b): a)
    choices_for_picking = [x for x in all_choices.values()]
    choices_for_picking.sort(reverse=True, key=lambda x: history.score(x))
    choice_id = frontend.pick(choices_for_picking)
    LOG.info("Chosen: %r", choice_id)
    if choice_id is not None:
        history.update(choice_id)
        command = "%s &" % all_choices[choice_id]['cmd']
        LOG.info("Running in shell: %r", command)
        subprocess.call(command, shell=True)


if __name__ == '__main__':
    main()
